jax/experimental/ode.py

```diff
@@ -12,343 +12,450 @@
+  Returns:
+    Initial step size for odeint algorithm.
+
+  Algorithm from:
+  E. Hairer, S. P. Norsett G. Wanner,
+  Solving Ordinary Differential Equations I: Nonstiff Problems, Sec. II.4.
+  """
+  scale = atol + np.abs(y0) * rtol
+  d0 = np.linalg.norm(y0 / scale)
+  d1 = np.linalg.norm(f0 / scale)
+  order_pow = (1. / (order + 1.))
+
+  h0 = jax.lax.cond(np.any(np.asarray([d0 < 1e-5, d1 < 1e-5])),
+                    1e-6,
+                    lambda x: x,
+                    0.01 * d0 / d1,
+                    lambda x: x)
+
+  y1 = y0 + h0 * f0
+  f1 = fun(y1, t0 + h0)
+  d2 = (np.linalg.norm(f1 - f0) / scale) / h0
+
+  h1 = jax.lax.cond(np.all(np.asarray([d0 <= 1e-15, d1 < 1e-15])),
+                    np.maximum(1e-6, h0 * 1e-3),
+                    lambda x: x,
+                    # order is always 4
+                    (0.01 / np.max(d1 + d2))**order_pow,
+                    lambda x: x)
+
+  return np.minimum(100. * h0, h1)
+
+
+@functools.partial(jax.jit, static_argnums=(0,))
 def runge_kutta_step(func, y0, f0, t0, dt):
-    """Take an arbitrary Runge-Kutta step and estimate error.
-    Args:
-        func: Function to evaluate like `func(t, y)` to compute the time derivative
-            of `y`.
-        y0: initial value for the state.
-        f0: initial value for the derivative, computed from `func(t0, y0)`.
-        t0: initial time.
-        dt: time step.
-        alpha, beta, c: Butcher tableau describing how to take the Runge-Kutta step.
-    Returns:
-        y1: estimated function at t1 = t0 + dt
-        f1: derivative of the state at t1
-        y1_error: estimated error at t1
-        k: list of Runge-Kutta coefficients `k` used for calculating these terms.
-    """
-    k = np.array([f0])
-    for alpha_i, beta_i in zip(alpha, beta):
-        ti = t0 + dt * alpha_i
-        yi = y0 + dt * np.dot(k.T, beta_i)
-        ft = func(yi, ti)
-        k = np.append(k, np.array([ft]), axis=0)
-
-    y1       = dt * np.dot(c_sol, k) + y0
-    y1_error = dt * np.dot(c_error, k)
-    f1 = k[-1]
-    return y1, f1, y1_error, k
-
-
-@jit
+  """Take an arbitrary Runge-Kutta step and estimate error.
+
+  Args:
+      func: Function to evaluate like `func(y, t)` to compute the time
+        derivative of `y`.
+      y0: initial value for the state.
+      f0: initial value for the derivative, computed from `func(t0, y0)`.
+      t0: initial time.
+      dt: time step.
+      alpha, beta, c: Butcher tableau describing how to take the Runge-Kutta
+        step.
+
+  Returns:
+      y1: estimated function at t1 = t0 + dt
+      f1: derivative of the state at t1
+      y1_error: estimated error at t1
+      k: list of Runge-Kutta coefficients `k` used for calculating these terms.
+  """
+  k = np.array([f0])
+  for alpha_i, beta_i in zip(alpha, beta):
+    ti = t0 + dt * alpha_i
+    yi = y0 + dt * np.dot(k.T, beta_i)
+    ft = func(yi, ti)
+    k = np.append(k, np.array([ft]), axis=0)
```

Comment:
It is better to use `np.concatenate` once (on a Python list of arrays) rather than to use `np.append` inside a loop, which has to reallocate the full result each time.